### ISSUE TITLE
Loses dynamically matched elements

### ISSUE THREAD

**Issue by dy** [WebReflection/regular-elements#21]
If we make element match selector dynamically, regular-elements/wicked-elements don't init that:

```js
import wels from "wicked-elements";

wels.define(".x", {
  init() {
    console.log("init");
  },
  onconnected() {
    console.log("connected");
  }
});

let el = document.createElement("div");
document.body.appendChild(el);

setTimeout(() => {
  el.classList.add("x");
}, 100);

```
[sandbox](https://codesandbox.io/s/busy-wozniak-ex6f5?fontsize=14&hidenavigation=1&theme=dark).

Direct mutation observer works fine
```js
var observer = new MutationObserver((mutationList, observer) => {
  console.log(...mutationList)
});
observer.observe(el, {childList: true, attributes: true, subtree: true})
document.body.appendChild(el);
```

**Comment by WebReflection** [comment]
Try this instead:
```js
import wels from "wicked-elements";
const $x = {
  init() {
    console.log("init");
  },
  onconnected() {
    console.log("connected");
  }
};
wels.define(".x", $x);

// further down ...
let el = document.createElement("div");
document.body.appendChild(el);

setTimeout(() => {
  el.classList.add("x");
  wels.define(el, $x);
}, 100);
```

The long story short, is that observing all attributes that could change on the DOM is a performance disaster, so since you are in charge of adding classes at runtime, you can as well use `.define(element, definition)` when that happens.

If you are not in charge, IIRC, the element will be upgraded whenever another element gets inserted, as _regularElements_ work on inserted nodes.

Accordingly, this looks like a won't fix so far, so please let me know if my proposed solution works.

**Comment by dy** [comment]
Got it. 

> performance disaster

I wonder if that's the case for CSS selectors as well - intuitively they provide similar reaction (rerendering a bunch of elements) on any DOM mutations, aren't them?

The lib I'm building ([spect](https://ghub.io/spect)) acts in a way similar to declarative set of rules, like CSS. I'd have to manually check all doc mutations then init regular elements, so that won't be performant solution either.

I think that's a question of benchmark.
For now I guess I have to opt for [selector-observer](https://ghub.io/selector-observer).

Thanks.

**Comment by WebReflection** [comment]
> I wonder if that's the case for CSS selectors as well

It doesn't matter, `MutationObserver` would trigger gazillion times more than it does when no attributes are observed, but it's true that filtering these by `class` attribute *could* be a decent compromise.

I'll re-open this, and keep an eye on noticeable performance regressions, as I don't think using another library would solve performance issues, since such library would be based on exactly same primitives.

**Comment by WebReflection** [comment]
P.S. do you also expect all nodes that don't match a selector anymore to be somehow freed from the logic? 'cause that's the biggest limitation I currently see in both regular/wickedElements, you can't drop an element from it's definition, so that once initialized, it'll keep doing whatever, if it doesn't match that selector anymore (but this might need a huge refactoring all over the place, so it might take a while).

**Comment by dy** [comment]
>  all nodes that don't match a selector anymore to be somehow freed from the logic

That's a good point.

I decided for elements initialized directly `$(realNodes, handler)` to permanently have assigned behavior, whereas for selectors `$('.behavior', handler)` to lose the behavior when elements are disconnected, basically the `ondisconnected` callback turns off the assigned aspect (same way the CSS would act).

~~I wonder if that's possible though - spect uses [augmentor](https://ghub.io/augmentor) as hooks provider by default, and there doesn't seem to be a way to dispose assigned hooks, is there?~~
Ah, [`dropEffect`](https://github.com/webreflection/augmentor#available-hooks), I see.

**Comment by WebReflection** [comment]
yes, `dropEffect` is used in [neverland](https://github.com/WebReflection/neverland/#readme) too, have a look in case it helps

**Comment by WebReflection** [comment]
Sorry, I meant [dom-augmentor](https://github.com/WebReflection/dom-augmentor/blob/master/esm/index.js), not _neverland_, but basically if you use _dom-augmentor_ instead of just _augmentor_, you'll have hooks pretty much 1:1 with React out of the box.

**Comment by WebReflection** [comment]
So, after thinking about it quite a while, but also prototyping a bit, I believe firing the whole circus per each possible attribute would make the library itself unusable, or better, a performance nightmare.

In details: _I cannot just listen to classes change_, for the simple reason that in CSS `[any-attr=value]` is a valid selector too.

As summary, whenever you want to create at runtime any reguler/wickedElement, you are in charge of also defining its behavior, otherwise you have to wait for any other node to be inserted on the page, so that the circus works out of the box.

As after years this runtime gotcha has came  up only once, I'll close this as won't fix for the mentioned reasons, and since you have a workaround that wouldn't kill this library performance for everyone else that never had such issue, I hope that's sensible/reasonable 👋
